crates/turbo-tasks-memory/src/edges_set.rs

```diff
@@ -0,0 +1,592 @@
+                            *self = EdgesEntry::OutputAndCell0(type_id);
+                            return true;
+                        }
+                        EdgesEntry::Child => {
+                            *self = EdgesEntry::ChildAndCell0(type_id);
+                            return true;
+                        }
+                        EdgesEntry::ChildAndOutput => {
+                            *self = EdgesEntry::ChildOutputAndCell0(type_id);
+                            return true;
+                        }
+                        _ => {}
+                    }
+                }
+            }
+            EdgeEntry::Collectibles(_) => {}
+        }
+        self.as_complex().insert(entry)
+    }
+
+    fn remove(&mut self, entry: EdgeEntry) -> bool {
+        if !self.has(entry) {
+            return false;
+        }
+        // We verified that the entry is present, so any non-complex case is easier to
+        // handle
+        match entry {
+            EdgeEntry::Output => match self {
+                EdgesEntry::Output => {
+                    *self = EdgesEntry::Empty;
+                    return true;
+                }
+                EdgesEntry::ChildAndOutput => {
+                    *self = EdgesEntry::Child;
+                    return true;
+                }
+                EdgesEntry::OutputAndCell0(type_id) => {
+                    *self = EdgesEntry::Cell0(*type_id);
+                    return true;
+                }
+                _ => {}
+            },
+            EdgeEntry::Child => match self {
+                EdgesEntry::Child => {
+                    *self = EdgesEntry::Empty;
+                    return true;
+                }
+                EdgesEntry::ChildAndOutput => {
+                    *self = EdgesEntry::Output;
+                    return true;
+                }
+                EdgesEntry::ChildAndCell0(type_id) => {
+                    *self = EdgesEntry::Cell0(*type_id);
+                    return true;
+                }
+                _ => {}
+            },
+            EdgeEntry::Cell(_) => match self {
+                EdgesEntry::Cell0(_) => {
+                    *self = EdgesEntry::Empty;
+                    return true;
+                }
+                EdgesEntry::OutputAndCell0(_) => {
+                    *self = EdgesEntry::Output;
+                    return true;
+                }
+                EdgesEntry::ChildAndCell0(_) => {
+                    *self = EdgesEntry::Child;
+                    return true;
+                }
+                _ => {}
+            },
+            EdgeEntry::Collectibles(_) => {}
+        }
+        if let EdgesEntry::Complex(set) = self {
+            if set.remove(&entry) {
+                self.simplify();
+                return true;
+            }
+        }
+        false
+    }
+
+    fn shrink_to_fit(&mut self) {
+        if let EdgesEntry::Complex(set) = self {
+            set.shrink_to_fit();
```

Comment:
If feel like this should also call simplify before trying to call `set.strink_to_fit`.

Reply:
Every operation that removes items already calls `simplify`, so we don't need to call it here.